### ISSUE TITLE
[CodeGen] Reserved registers frozen multiple times

### ISSUE THREAD

**Issue by nikic** [llvm/llvm-project#71692]
Based on our documentation, it sounds like the set of ReservedRegs is supposed to stay constant after it has been frozen:
https://github.com/llvm/llvm-project/blob/98d8b688bd351850ca99d3e5e9dfdf582c83c4f4/llvm/include/llvm/CodeGen/MachineRegisterInfo.h#L901-L911

However, in practice we call freezeReservedRegs() at least twice, once during FinalizeISel and then again in RegAlloc. Adding a check to not touch the ReservedRegs if they have already been computed results in a bunch of verifier failures in AMDGPU.

I think *something* is wrong here (possibly just our documentation), but I'm not familiar enough with the subject matter to figure out what...

**Comment by jayfoad** [comment]
Yeah, I believe targets are allowed to update the set of frozen registers, but I don't know what the rules are exactly. 
See also https://reviews.llvm.org/D119436 "RFC: RegAlloc: Avoid repeated calls to freezeReservedRegs"

**Comment by nikic** [comment]
> Yeah, I believe targets are allowed to update the set of frozen registers, but I don't know what the rules are exactly. See also https://reviews.llvm.org/D119436 "RFC: RegAlloc: Avoid repeated calls to freezeReservedRegs"

Oooh, that patch is exactly what I wanted to see! Looks like this isn't as simple as one would have hoped.

**Comment by arsenm** [comment]
The documentation is overly strict. Really the constraint is you can't change the set of reserved registers mid-allocation. More broadly, if you have a working data structure tracking the liveness. Similarly you shouldn't change it while you have a RegisterScavenger tracking

**Comment by CRobeck** [comment]
I remember running into this when reserving registers for long branches. We set an initial set of reserved registers and then compact down to the lowest unused ones. At the time it wasn't clear when the latest point we could do that was. It would be good to consolidate that. reserveReg is similar freezeReservedRegs but less of a hammer in terms of recomputing.. That section of code is here:

https://github.com/llvm/llvm-project/blob/11ac97c67a9315dce48fd938d68ae991e3559f10/llvm/lib/Target/AMDGPU/SIFrameLowering.cpp#L1458-L1484
